kapitan/refs/secrets/vault.py

```diff
@@ -0,0 +1,229 @@
+# Copyright 2019 The Kapitan Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"hashicorp vault secrets module"
+
+import hvac
+import base64
+from binascii import Error as b_error
+from hvac.exceptions import Forbidden, VaultError, InvalidPath
+from yaml import safe_load
+from os.path import join,expanduser
+from os import getenv
+from sys import argv, exit
+
+from kapitan.refs.base import Ref, RefBackend, RefError
+from kapitan import cached
+from kapitan.errors import KapitanError
+
+class VaultError(KapitanError):
+    """Generic vault errors"""
+    pass
+
+def get_env(parameter):
+    """
+    The following variables need to be exported to the environment depending on authentication needed
+    * VAULT_ADDR: url for vault
+    * VAULT_SKIP_VERIFY=true: if set, do not verify presented TLS certificate before communicating with Vault server.
+    * VAULT_TOKEN: token for vault
+    * VAULT_ROLE_ID: (required by approle)
+    * VAULT_SECRET_ID: (required by approle)
+    * VAULT_USERNAME: username to login to vault
+    * VAULT_PASSWORD: password to login to vault
+    * VAULT_CLIENT_KEY: path to an unencrypted PEM-encoded private key matching the client certificate
+    * VAULT_CLIENT_CERT: path to a PEM-encoded client certificate for TLS authentication to the Vault server
+    * VAULT_CACERT: path to a PEM-encoded CA cert file to use to verify the Vault server TLS certificate
+    * VAULT_CAPATH: path to a directory of PEM-encoded CA cert files to verify the Vault server TLS certificate
+    * VAULT_NAMESPACE: specify the Vault Namespace, if you have one
+    """
+    # Helper lambda function to get varibles either from parameters or environment
+    get_variable = lambda x: parameter.get(x, getenv('VAULT_'+x.upper(),default=''))
+    env = {}
+    env['url'] = get_variable('addr')
+    env['namespace'] = get_variable('namespace')
+    # AUTHENTICATION TYPE
+    if parameter.get('auth','token') in ['token','github']:
+        env['token'] = getenv('VAULT_TOKEN')
```

Comment:
Using get_variable to grab vault token would be inappropriate here since the token would be either in the environment or in `~/.vault-token` but never in the inventory.